operator/pkg/helmreconciler/apply.go

```diff
@@ -92,8 +100,21 @@ func (h *HelmReconciler) ApplyManifest(manifest name.Manifest, serverSideApply b
 	// Objects are applied in groups: namespaces, CRDs, everything else, with wait for ready in between.
 	nsObjs := object.KindObjects(changedObjects, name.NamespaceStr)
 	crdObjs := object.KindObjects(changedObjects, name.CRDStr)
-	otherObjs := object.ObjectsNotInLists(changedObjects, nsObjs, crdObjs)
-	for _, objList := range []object.K8sObjects{nsObjs, crdObjs, otherObjs} {
+	mutatingWebhookObjs := object.KindObjects(changedObjects, name.MutatingWebhookConfigurationStr)
+	otherObjs := object.ObjectsNotInLists(changedObjects, nsObjs, crdObjs, mutatingWebhookObjs)
+	for _, objList := range []object.K8sObjects{nsObjs, crdObjs, otherObjs, mutatingWebhookObjs} {
+		// Analyze mutating webhook objects
+		if len(objList) > 0 {
+			if objList[0].Kind == name.MutatingWebhookConfigurationStr {
+				err := h.analyzeWebhooks(objList)
```

Comment:
One issue with this is the order:
```
$ /istioctl/cmd/istioctl install -y
✔ Istio core installed
- Processing resources for Istiod.                                                                                                                                                    2021-06-28T17:27:09.163949Z     error   installer       creating default tag would conflict:
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/namespace.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/object.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/rev.namespace.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/rev.object.sidecar-injector.istio.io]. This may cause injection to occur twice.
✘ Istiod encountered an error: creating default tag would conflict:
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/namespace.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/object.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/rev.namespace.sidecar-injector.istio.io]. This may cause injection to occur twice.
Error [IST0139] (MutatingWebhookConfiguration istio-sidecar-injector ) Webhook overlaps with others: [istio-revision-tag-default/rev.object.sidecar-injector.istio.io]. This may cause injection to occur twice.
✔ Ingress gateways installed
- Pruning removed resources                                                                                                                                                           Error: failed to install manifests: errors occurred during operation
```

So the install still happens. Ideally we do it upfront, and fail without writing anything to the cluster

Reply:
I actually wanted some insights on this, Initially I had implemented what you have mentioned.
The issue with that is, Analyzer also looks for a valid istiod service running in the cluster. If validate webhooks upfront it will always fail and happy installation path never goes through.

Reply:
Since, The installation path is just like kubectl apply it doesn't add unnecessary churn
If user fixes the issue with the webhook and run the install command, Only the webhook object will be created and all else remains the same.
So, This was like a clean and stable solution I could think of like just skipping the webhook creation on error and letting everything else go through.

Reply:
Since, The installation path is just like kubectl apply it doesn't add unnecessary churn
If user fixes the issue with the webhook and run the install command, Only the webhook object will be created and all else remains the same.
So, This was like a clean and stable solution I could think of like just skipping the webhook creation on error and letting everything else go through.

Reply:
@howardjohn 
Should we think on a solution for this?
Or partial install works?

Reply:
I would prefer to fail (or prompt?) rather than partial install. Generally it seems favorable to have atomic operations, so everything succeeds or nothing does.

cc @Monkeyanator if you have opinions here

Reply:
I agree that ideally we should fail rather than continue with partial install. Istio is loud on install and error messages can be easy to ignore IMO

Reply:
I have updated the PR
Now, we remove anything installed in case there are some errors and --force is not set

Reply:
I would much prefer we do an upfront check. Adding things then removing them may lead to unexpected behavior (example: I already have istio install, and I rerun `istioctl install`. Now it will remove my original installation).

Reply:
@howardjohn 
Issue with upfront validation is that it will always fail because webhook analyzer requires Istiod/Pilot component corresponding to the webhook being analysed installed.
So the upfront validation will always fail.

To acknowledge your concern, If you rerun istioctl install will not remove original installation as there will be no changedObjects to be deployed
But in case istioctl install --revision=somevalue is given and it results in some conflicts in this case we might run into some errors.
I think what we should do is instead of removing allObjects we should just remove the changedObjects that should be fine, WDYT?

Reply:
Or we can skip the part in the webhook analyser where it checks for the presence of Istiod service for the `istioctl install` workflow

Reply:
if we pass all the yaml we are about to apply (including the service) would that work?

if not, I would prefer to skip that step.

Thanks for iterating on this?

Reply:
@howardjohn 
I have updated the workflow to do the analysis upfront before creating anything on the cluster
If force is not set installation will exit with an error without creating/deleting any object from the cluster else normal flow will continue if force is set